telemetry/convert.py

```diff
@@ -186,11 +269,67 @@ def convert_obj(self, json_dict, date, ip=None):
+
+            # Verify some basic assumptions
+            if (pingVersion != Converter.VERSION_UNIFIED and
+                pingVersion != Converter.VERSION_UNIFIED_CONVERTED):
+                raise ValueError("Unknown unified ping version: " + str(json_dict["version"]))
+            elif not pingType:
+                raise ValueError("Unified ping has no top-level 'type' field")
+            elif "application" not in json_dict:
+                raise ValueError("Unified ping has no top-level 'application' field")
+            elif "payload" not in json_dict:
+                raise ValueError("Unified ping has no top-level 'payload' field")
+
+            if pingVersion == Converter.VERSION_UNIFIED:
+                # Create a ping.info field and copy select ping.application fields.
+                # This is the minimum required for compatibility with the calling
+                # process_incoming* scripts
+                if "info" in json_dict:
+                    raise ValueError("Raw unified ping has an existing ping.info field")
+
+                info = json_dict["info"] = {}
+                self.add_info_fields(info, json_dict["application"], appFieldMap)
+                info["reason"] = pingType
+                json_dict["version"] = Converter.VERSION_UNIFIED_CONVERTED
+
+                if pingType == "saved-session" or pingType == "main":
+                    # This is a unified ping in the "main" ping format:
+                    # https://ci.mozilla.org/job/mozilla-central-docs/Tree_Documentation/toolkit/components/telemetry/telemetry/main-ping.html
+                    json_payload = json_dict["payload"] 
+                    if "info" not in json_payload:
+                        raise ValueError("Unified " + pingType + " missing payload.info")
+                    elif "histograms" not in json_payload:
+                        raise ValueError("Unified " + pingType + " missing payload.histograms")
+                    # Convert the histograms section to the more compact representation
+                    self.convert_histograms(json_payload)
+
+                    # Make unified saved-session pings look like classic pings
+                    if pingType == "saved-session":
+                        if not "environment" in json_dict:
+                            raise ValueError("Unified saved-session ping missing 'environment' field")
+                        self.convert_saved_session(json_dict)
+                else:
+                    # Future ping type (activation, upgrade, deletion)
+                    # Don't alter it further
+                    pass
+            else:
+                # pingVersion == Converter.VERSION_UNIFIED_CONVERTED:
+                # This unified ping was already converted. Do nothing
+                info = json_dict["info"]
+                pass
         else:
-            raise ValueError("Missing payload version")
+            raise ValueError("Missing ping.ver and ping.version")
```

Comment:
This string is used by the (admittedly weak) tests - please update `telemetry/test_convert.py` accordingly